File: sdks/ts/memory/src/memory/reflect.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Reflect stage. End-of-session pass that asks the provider to extract
 * generalisable heuristics + a short summary, then persists a reflection
 * file at `reflections/<sessionId>.md` with structured frontmatter + open
 * questions. Writes flow through a single Store batch so partial
 * reflections cannot leak out on error.
 */

import type { Logger, Message, Provider } from '../llm/index.js'
import type { Store } from '../store/index.js'
import { buildFrontmatter } from './frontmatter.js'
import { parseFrontmatter } from './frontmatter.js'
import { reflectionPath, scopeIndex, scopeTopic } from './paths.js'
import { fireReflectionEnd, fireReflectionStart } from './plugins.js'
import { REFLECTION_SYSTEM_PROMPT } from './prompts.js'
import type {
  Heuristic,
  Plugin,
  ReflectArgs,
  ReflectionResult,
  Scope,
} from './types.js'

const REFLECT_MAX_TOKENS = 4096
const REFLECT_TEMPERATURE = 0.3
const REFLECT_MAX_RECENT = 60
const HEURISTIC_TAG = 'heuristic'
const HEURISTIC_ENTRY_LIMIT = 140

export type ReflectDeps = {
  readonly store: Store
  readonly provider: Provider
  readonly logger: Logger
  readonly plugins: readonly Plugin[]
  readonly defaultScope: Scope
  readonly defaultActorId: string
}

export const createReflect = (deps: ReflectDeps) => {
  return async (args: ReflectArgs): Promise<ReflectionResult | undefined> => {
    const actorId = args.actorId ?? deps.defaultActorId
    const scope = args.scope ?? deps.defaultScope
    const messages = args.messages

    await fireReflectionStart(
      deps.plugins,
      { actorId, scope, messages },
      deps.logger,
    )

    const recent =
      messages.length > REFLECT_MAX_RECENT
        ? messages.slice(-REFLECT_MAX_RECENT)
        : messages
    const userPrompt = buildReflectionPrompt(recent)

    let raw: string
    try {
      const resp = await deps.provider.complete({
        messages: [{ role: 'user', content: userPrompt }],
        system: REFLECTION_SYSTEM_PROMPT,
        maxTokens: REFLECT_MAX_TOKENS,
        temperature: REFLECT_TEMPERATURE,
      })
      raw = resp.content
    } catch (err) {
      deps.logger.warn('memory: reflect provider call failed', {
        err: err instanceof Error ? err.message : String(err),
      })
      return undefined
    }

    const parsed = parseReflectionJson(raw)
    if (!parsed.outcome) {
      await fireReflectionEnd(
        deps.plugins,
        { actorId, scope, messages },
        deps.logger,
      )
      return undefined
    }

    const path = reflectionPath(args.sessionId)
    const fileContent = buildReflectionFile(parsed, args.sessionId, scope)

    try {
      await deps.store.batch({ reason: 'reflect' }, async (b) => {
        await b.write(path, Buffer.from(fileContent, 'utf8'))
        await persistHeuristicsInBatch(b, {
          actorId,
          sessionId: args.sessionId,
          heuristics: parsed.heuristics,
        })
      })
    } catch (err) {
      deps.logger.warn('memory: reflect persist failed', {
        err: err instanceof Error ? err.message : String(err),
      })
      return undefined
    }

    const result: ReflectionResult = { ...parsed, path }
    await fireReflectionEnd(
      deps.plugins,
      { actorId, scope, messages, result },
      deps.logger,
    )
    return result
  }
}

type ParsedReflection = {
  outcome: 'success' | 'partial' | 'failure' | 'unknown'
  summary: string
  retryFeedback: string
  shouldRecordEpisode: boolean
  openQuestions: readonly string[]
  heuristics: readonly Heuristic[]
}

export const parseReflectionJson = (content: string): ParsedReflection => {
  const trimmed = content.trim()
  const first = trimmed.indexOf('{')
  const last = trimmed.lastIndexOf('}')
  if (first < 0 || last <= first) {
    return {
      outcome: 'unknown',
      summary: '',
      retryFeedback: '',
      shouldRecordEpisode: false,
      openQuestions: [],
      heuristics: [],
    }
  }
  const slice = trimmed.slice(first, last + 1)
  try {
    const parsed = JSON.parse(slice) as Record<string, unknown>
    const outcome =
      parsed.outcome === 'success' ||
      parsed.outcome === 'partial' ||
      parsed.outcome === 'failure'
        ? parsed.outcome
        : 'unknown'
    const summary = typeof parsed.summary === 'string' ? parsed.summary : ''
    const retryFeedback =
      typeof parsed.retry_feedback === 'string'
        ? parsed.retry_feedback
        : typeof parsed.retryFeedback === 'string'
          ? parsed.retryFeedback
          : ''
    const shouldRecordEpisode =
      typeof parsed.should_record_episode === 'boolean'
        ? parsed.should_record_episode
        : typeof parsed.shouldRecordEpisode === 'boolean'
          ? parsed.shouldRecordEpisode
          : false
    const openQuestionsRaw = Array.isArray(parsed.open_questions)
      ? parsed.open_questions
      : Array.isArray(parsed.openQuestions)
        ? parsed.openQuestions
        : []
    const openQuestions = openQuestionsRaw.filter(
      (s): s is string => typeof s === 'string',
    )
    const heuristicsRaw = Array.isArray(parsed.heuristics) ? parsed.heuristics : []
    const heuristics: Heuristic[] = []
    for (const h of heuristicsRaw) {
      if (typeof h !== 'object' || h === null) continue
      const r = h as Record<string, unknown>
      heuristics.push({
        rule: typeof r.rule === 'string' ? r.rule : '',
        context: typeof r.context === 'string' ? r.context : '',
        confidence:
          r.confidence === 'high' || r.confidence === 'medium' || r.confidence === 'low'
            ? r.confidence
            : 'low',
        category: typeof r.category === 'string' ? r.category : '',
        scope:
          r.scope === 'global' || r.scope === 'project' || r.scope === 'agent'
            ? r.scope
            : 'project',
        antiPattern:
          r.anti_pattern === true || r.antiPattern === true ? true : false,
      })
    }
    return {
      outcome,
      summary,
      retryFeedback,
      shouldRecordEpisode,
      openQuestions,
      heuristics,
    }
  } catch {
    return {
      outcome: 'unknown',
      summary: '',
      retryFeedback: '',
      shouldRecordEpisode: false,
      openQuestions: [],
      heuristics: [],
    }
  }
}

const buildReflectionPrompt = (messages: readonly Message[]): string => {
  const parts: string[] = ['## Session transcript', '']
  for (const m of messages) {
    let content = m.content ?? ''
    if (content.length > 1000) content = `${content.slice(0, 1000)}\n[...truncated]`
    if (m.role === 'tool') {
      parts.push(`[tool ${m.name ?? ''}]: ${content}`, '')
      continue
    }
    parts.push(`[${m.role}]: ${content}`, '')
  }
  return parts.join('\n')
}

const buildReflectionFile = (
  r: ParsedReflection,
  sessionId: string,
  scope: Scope,
): string => {
  const now = new Date().toISOString()
  const fm = buildFrontmatter({
    extra: {
      should_record_episode: r.shouldRecordEpisode ? 'true' : 'false',
    },
    type: 'reflection',
    scope,
    created: now,
    modified: now,
    source: 'reflection',
    session_id: sessionId,
  })
  const body: string[] = []
  body.push(`# Session ${sessionId}`)
  body.push('')
  body.push(`**Outcome:** ${r.outcome}`)
  body.push(`**Record episode:** ${r.shouldRecordEpisode ? 'yes' : 'no'}`)
  body.push('')
  body.push('## Summary', '', r.summary || '_no summary_', '')
  body.push('## Retry feedback', '', r.retryFeedback || '_none_', '')
  if (r.openQuestions.length > 0) {
    body.push('## Open questions', '')
    for (const q of r.openQuestions) body.push(`- ${q}`)
    body.push('')
  }
  if (r.heuristics.length > 0) {
    body.push('## Heuristics', '')
    for (const h of r.heuristics) {
      const marker = h.antiPattern ? '[anti-pattern]' : '[pattern]'
      body.push(
        `- ${marker} ${h.rule} _(context: ${h.context}; confidence: ${h.confidence}; category: ${h.category}; scope: ${h.scope})_`,
      )
    }
    body.push('')
  }
  return `${fm}\n${body.join('\n')}`
}

const persistHeuristicsInBatch = async (
  batch: import('../store/index.js').Batch,
  input: {
    readonly actorId: string
    readonly sessionId: string
    readonly heuristics: readonly Heuristic[]
  },
): Promise<void> => {
  if (input.heuristics.length === 0) return

  const groupedEntries = new Map<Scope, string[]>()
  const now = new Date().toISOString()
  const deduped = new Map<string, Heuristic>()
  for (const heuristic of input.heuristics) {
    deduped.set(heuristicStoragePath(heuristic, input.actorId), heuristic)
  }

  for (const [path, heuristic] of deduped) {
    const existing = await batch.read(path as import('../store/path.js').Path).catch(() => undefined)
    const parsedExisting =
      existing === undefined ? undefined : parseFrontmatter(existing.toString('utf8')).frontmatter
    await batch.write(
      path as import('../store/path.js').Path,
      Buffer.from(
        buildHeuristicFile({
          heuristic,
          sessionId: input.sessionId,
          created: parsedExisting?.created,
          now,
        }),
        'utf8',
      ),
    )
    const entries = groupedEntries.get(heuristic.scope) ?? []
    entries.push(heuristicIndexEntry(path, heuristic))
    groupedEntries.set(heuristic.scope, entries)
  }

  for (const [scope, entries] of groupedEntries) {
    await appendIndexEntries(batch, scopeIndex(scope, input.actorId), entries)
  }
}

const appendIndexEntries = async (
  batch: import('../store/index.js').Batch,
  indexPath: import('../store/path.js').Path,
  entries: readonly string[],
): Promise<void> => {
  let existing = ''
  try {
    existing = (await batch.read(indexPath)).toString('utf8').trim()
  } catch {
    // no index yet
  }
  let content = existing
  for (const raw of entries) {
    const entry = raw.trim()
    if (entry === '' || content.includes(entry)) continue
    content = content === '' ? entry : `${content}\n${entry}`
  }
  await batch.write(indexPath, Buffer.from(`${content}\n`, 'utf8'))
}

const heuristicStoragePath = (heuristic: Heuristic, actorId: string): string =>
  scopeTopic(heuristic.scope, actorId, heuristicFilename(heuristic))

const heuristicFilename = (heuristic: Heuristic): string => {
  const prefix = heuristic.antiPattern ? 'anti-pattern' : 'heuristic'
  return `${prefix}-${slugify(heuristic.category)}-${slugify(heuristic.rule)}.md`
}

const heuristicIndexEntry = (path: string, heuristic: Heuristic): string =>
  truncateOneLine(
    `- ${path.split('/').pop() ?? path}: ${(heuristic.antiPattern ? 'anti-pattern' : 'heuristic')} ${heuristic.rule}`,
    HEURISTIC_ENTRY_LIMIT,
  )

const buildHeuristicFile = (input: {
  readonly heuristic: Heuristic
  readonly sessionId: string
  readonly created?: string | undefined
  readonly now: string
}): string => {
  const { heuristic } = input
  const fm = buildFrontmatter({
    extra: {},
    name: heuristic.antiPattern ? `Anti-pattern: ${heuristic.rule}` : `Heuristic: ${heuristic.rule}`,
    description: heuristicDescription(heuristic),
    type: heuristic.scope === 'project' ? 'project' : 'feedback',
    scope: heuristic.scope,
    created: input.created ?? input.now,
    modified: input.now,
    confidence: heuristic.confidence,
    source: 'reflection',
    session_id: input.sessionId,
    tags: heuristicTags(heuristic),
  })
  const body = heuristic.antiPattern
    ? [
        `Anti-pattern: ${heuristic.rule}`,
        `Context: ${heuristicContext(heuristic)}`,
        'Why: This caused friction in a previous reflected session.',
        `How to apply: Avoid this when ${heuristicContext(heuristic)}.`,
      ]
    : [
        `Rule: ${heuristic.rule}`,
        `Context: ${heuristicContext(heuristic)}`,
        'Why: This pattern helped in a previous reflected session.',
        `How to apply: Use this when ${heuristicContext(heuristic)}.`,
      ]
  return `${fm}\n${body.join('\n\n')}\n`
}

const heuristicDescription = (heuristic: Heuristic): string =>
  truncateOneLine(
    `${heuristic.antiPattern ? 'Anti-pattern' : 'Heuristic'} for ${heuristicContext(heuristic)}`,
    140,
  )

const heuristicTags = (heuristic: Heuristic): readonly string[] =>
  [
    HEURISTIC_TAG,
    heuristic.category.trim().toLowerCase(),
    heuristic.confidence.trim().toLowerCase(),
    ...(heuristic.antiPattern ? ['anti-pattern'] : ['pattern']),
  ].filter((tag, index, tags) => tag !== '' && tags.indexOf(tag) === index)

const heuristicContext = (heuristic: Heuristic): string =>
  heuristic.context.trim() !== '' ? heuristic.context.trim() : 'the same type of work'

const slugify = (value: string): string => {
  const slug = value
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '')
  return slug === '' ? 'note' : slug.slice(0, 64).replace(/-+$/g, '')
}

const truncateOneLine = (value: string, limit: number): string => {
  const collapsed = value.replace(/\s+/g, ' ').trim()
  if (collapsed.length <= limit) return collapsed
  return `${collapsed.slice(0, Math.max(limit - 3, 1)).trimEnd()}...`
}

File: go/memory/reflect.go
// SPDX-License-Identifier: Apache-2.0

package memory

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"strings"
	"sync"

	"github.com/jeffs-brain/memory/go/llm"
)

// Reflection thresholds.
const (
	reflectMaxTokens   = 4096
	reflectTemperature = 0.3
	reflectMinMessages = 8
	reflectMaxRecent   = 60
)

// Reflector manages background session reflection.
type Reflector struct {
	mem *Memory

	mu         sync.Mutex
	lastCursor int
	inProgress bool
}

// NewReflector creates a new Reflector bound to the supplied Memory.
func NewReflector(mem *Memory) *Reflector {
	return &Reflector{mem: mem}
}

// ReflectionResult holds the parsed output from the reflection agent.
type ReflectionResult struct {
	Outcome             string      `json:"outcome"`
	Summary             string      `json:"summary"`
	RetryFeedback       string      `json:"retry_feedback"`
	Heuristics          []Heuristic `json:"heuristics"`
	ShouldRecordEpisode bool        `json:"should_record_episode"`
}

// Heuristic represents a single generalisable pattern extracted from a
// session.
type Heuristic struct {
	Rule        string `json:"rule"`
	Context     string `json:"context"`
	Confidence  string `json:"confidence"`
	Category    string `json:"category"`
	Scope       string `json:"scope"`
	AntiPattern bool   `json:"anti_pattern"`
}

// SessionAnalysis summarises a coding session for the reflection
// prompt.
type SessionAnalysis struct {
	TaskDescription   string
	ToolCallSummary   string
	ErrorsEncountered []string
	UserCorrections   []string
	IterationCount    int
	WriteToolCalls    int
	Outcome           string
}

// reflectionSystemPrompt drives generalisation, not memorisation.
// Ported verbatim from jeff.
const reflectionSystemPrompt = `You are a reflection agent. You analyse completed coding sessions to extract lasting wisdom.

Your job is NOT to summarise what happened — it is to identify GENERALISABLE PATTERNS.

Good heuristic: "When working on Go projects with generated code, check for //go:generate directives before modifying generated files."
Bad heuristic: "The file cmd/server/main.go has a bug on line 42." (Too specific.)

## Output format
Respond with ONLY valid JSON:
{
  "outcome": "success|partial|failure",
  "summary": "one paragraph",
  "retry_feedback": "what to do differently if retrying this specific task",
  "heuristics": [
    {
      "rule": "imperative, actionable pattern",
      "context": "when this applies (language, framework, problem type)",
      "confidence": "low|medium|high",
      "category": "approach|debugging|architecture|testing|communication",
      "scope": "project|global",
      "anti_pattern": false
    }
  ],
  "should_record_episode": true
}

## When to produce heuristics
- User corrected the agent → HIGH confidence (possibly anti_pattern=true)
- Multiple approaches tried before success → MEDIUM confidence
- Non-obvious error encountered → LOW confidence
- Routine session → empty array is fine

## Anti-pattern signals
Look for: "no", "don't", "stop", "instead", "that's wrong", "not like that", agent backtracking, multiple failed attempts`

// correctionPatterns are phrases that signal the user corrected the
// agent.
var correctionPatterns = []string{
	"no,", "no ", "don't", "do not", "stop", "instead",
	"that's wrong", "not like that", "not what i",
	"please revert", "undo that", "try again",
}

// MaybeReflect checks if reflection should run and, if so, analyses
// the session to extract generalisable heuristics.
func (r *Reflector) MaybeReflect(
	ctx context.Context,
	provider llm.Provider,
	model string,
	projectPath string,
	messages []Message,
) {
	r.mu.Lock()
	if r.inProgress {
		r.mu.Unlock()
		return
	}
	r.inProgress = true
	cursor := r.lastCursor
	r.mu.Unlock()

	defer func() {
		r.mu.Lock()
		r.inProgress = false
		r.mu.Unlock()
	}()

	if !shouldReflect(messages, cursor) {
		return
	}

	result := r.reflect(ctx, provider, model, projectPath, messages, cursor)
	if result == nil {
		return
	}

	if len(result.Heuristics) > 0 {
		if err := r.mem.ApplyHeuristics(ctx, ProjectSlug(projectPath), result.Heuristics); err != nil {
			slog.Warn("memory: apply heuristics failed", "err", err)
		}
	}

	r.mu.Lock()
	r.lastCursor = len(messages)
	r.mu.Unlock()
}

// ForceReflect skips the shouldReflect check and returns the result
// directly.
func (r *Reflector) ForceReflect(
	ctx context.Context,
	provider llm.Provider,
	model string,
	projectPath string,
	messages []Message,
) *ReflectionResult {
	r.mu.Lock()
	cursor := r.lastCursor
	r.mu.Unlock()

	result := r.reflect(ctx, provider, model, projectPath, messages, cursor)
	if result == nil {
		return nil
	}

	if len(result.Heuristics) > 0 {
		if err := r.mem.ApplyHeuristics(ctx, ProjectSlug(projectPath), result.Heuristics); err != nil {
			slog.Warn("memory: apply heuristics failed", "err", err)
		}
	}

	r.mu.Lock()
	r.lastCursor = len(messages)
	r.mu.Unlock()

	return result
}

// reflect performs the actual LLM call and result parsing.
func (r *Reflector) reflect(
	ctx context.Context,
	provider llm.Provider,
	model string,
	projectPath string,
	messages []Message,
	cursor int,
) *ReflectionResult {
	recent := messages[cursor:]
	if len(recent) > reflectMaxRecent {
		recent = recent[len(recent)-reflectMaxRecent:]
	}

	analysis := analyseSession(recent)
	userPrompt := buildReflectionPrompt(analysis, recent)

	resp, err := provider.Complete(ctx, llm.CompleteRequest{
		Model: model,
		Messages: []llm.Message{
			{Role: RoleSystem, Content: reflectionSystemPrompt},
			{Role: RoleUser, Content: userPrompt},
		},
		MaxTokens:   reflectMaxTokens,
		Temperature: reflectTemperature,
	})
	if err != nil {
		return nil
	}

	result := parseReflectionResult(resp.Text)
	if result.Outcome == "" {
		return nil
	}

	return &result
}

// shouldReflect determines whether the session warrants reflection.
func shouldReflect(messages []Message, cursor int) bool {
	newMessages := messages[cursor:]

	if len(newMessages) < reflectMinMessages {
		return false
	}

	writes := countWriteToolCalls(newMessages)
	corrections := findUserCorrections(newMessages)
	iterations := countAssistantToolIterations(newMessages)

	if writes == 0 && iterations < 10 && len(corrections) == 0 {
		return false
	}

	return true
}

// analyseSession builds a structured summary of the session.
func analyseSession(messages []Message) SessionAnalysis {
	return SessionAnalysis{
		TaskDescription:   extractTaskDescription(messages),
		ToolCallSummary:   summariseToolCalls(messages),
		ErrorsEncountered: findErrors(messages),
		UserCorrections:   findUserCorrections(messages),
		IterationCount:    countAssistantToolIterations(messages),
		WriteToolCalls:    countWriteToolCalls(messages),
		Outcome:           inferOutcome(messages),
	}
}

// extractTaskDescription returns the first substantive user message as
// a proxy for the task description.
func extractTaskDescription(messages []Message) string {
	for _, m := range messages {
		if m.Role != RoleUser {
			continue
		}
		content := strings.TrimSpace(m.Content)
		if len(content) < 5 {
			continue
		}
		if len(content) > 500 {
			content = content[:500] + "..."
		}
		return content
	}
	return ""
}

// summariseToolCalls groups tool calls by name and extracts file paths
// for write/edit operations.
func summariseToolCalls(messages []Message) string {
	type toolInfo struct {
		count int
		files []string
	}
	tools := make(map[string]*toolInfo)
	var order []string

	for _, m := range messages {
		if m.Role != RoleAssistant {
			continue
		}
		for _, tc := range m.ToolCalls {
			name := tc.Name
			info, ok := tools[name]
			if !ok {
				info = &toolInfo{}
				tools[name] = info
				order = append(order, name)
			}
			info.count++

			if name == "write" || name == "edit" {
				if path := extractFilePath(tc.Arguments); path != "" {
					info.files = append(info.files, path)
				}
			}
		}
	}

	var b strings.Builder
	for _, name := range order {
		info := tools[name]
		b.WriteString(fmt.Sprintf("%s: %d calls", name, info.count))
		if len(info.files) > 0 {
			seen := make(map[string]bool)
			var unique []string
			for _, f := range info.files {
				base := baseFilename(f)
				if !seen[base] {
					seen[base] = true
					unique = append(unique, base)
				}
			}
			b.WriteString(fmt.Sprintf(" (%s)", strings.Join(unique, ", ")))
		}
		b.WriteString("\n")
	}

	return strings.TrimSpace(b.String())
}

// extractFilePath attempts to pull a file_path from JSON tool
// arguments.
func extractFilePath(args json.RawMessage) string {
	var parsed struct {
		FilePath string `json:"file_path"`
	}
	if err := json.Unmarshal(args, &parsed); err != nil {
		return ""
	}
	return parsed.FilePath
}

// baseFilename returns the last component of a file path.
func baseFilename(path string) string {
	parts := strings.Split(path, "/")
	if len(parts) == 0 {
		return path
	}
	return parts[len(parts)-1]
}

// findUserCorrections scans user messages for correction patterns.
func findUserCorrections(messages []Message) []string {
	var corrections []string
	for _, m := range messages {
		if m.Role != RoleUser {
			continue
		}
		lower := strings.ToLower(m.Content)
		for _, pattern := range correctionPatterns {
			if strings.Contains(lower, pattern) {
				content := m.Content
				if len(content) > 200 {
					content = content[:200] + "..."
				}
				corrections = append(corrections, content)
				break
			}
		}
	}
	return corrections
}

// findErrors scans tool results for error indicators.
func findErrors(messages []Message) []string {
	var errors []string
	for _, m := range messages {
		if m.Role != RoleTool {
			continue
		}
		lower := strings.ToLower(m.Content)
		if strings.Contains(lower, "error") || strings.Contains(lower, "failed") ||
			strings.Contains(lower, "panic") || strings.Contains(lower, "cannot") ||
			strings.Contains(lower, "fatal") {
			content := m.Content
			if len(content) > 300 {
				content = content[:300] + "..."
			}
			errors = append(errors, content)
		}

		for _, block := range m.Blocks {
			if block.ToolResult != nil && block.ToolResult.IsError {
				content := block.ToolResult.Content
				if len(content) > 300 {
					content = content[:300] + "..."
				}
				errors = append(errors, content)
			}
		}
	}
	return errors
}

// countWriteToolCalls counts write and edit tool calls in the
// messages.
func countWriteToolCalls(messages []Message) int {
	count := 0
	for _, m := range messages {
		if m.Role != RoleAssistant {
			continue
		}
		for _, tc := range m.ToolCalls {
			if tc.Name == "write" || tc.Name == "edit" {
				count++
			}
		}
	}
	return count
}

// countAssistantToolIterations counts assistant messages that contain
// tool calls.
func countAssistantToolIterations(messages []Message) int {
	count := 0
	for _, m := range messages {
		if m.Role == RoleAssistant && len(m.ToolCalls) > 0 {
			count++
		}
	}
	return count
}

// inferOutcome guesses the session outcome from the final messages.
func inferOutcome(messages []Message) string {
	if len(messages) == 0 {
		return "unknown"
	}

	end := len(messages)
	start := end - 5
	if start < 0 {
		start = 0
	}

	for i := end - 1; i >= start; i-- {
		m := messages[i]
		lower := strings.ToLower(m.Content)

		if m.Role == RoleUser {
			if strings.Contains(lower, "thanks") || strings.Contains(lower, "perfect") ||
				strings.Contains(lower, "great") || strings.Contains(lower, "looks good") {
				return "success"
			}
			if strings.Contains(lower, "that's wrong") || strings.Contains(lower, "not working") ||
				strings.Contains(lower, "broken") || strings.Contains(lower, "revert") {
				return "failure"
			}
		}
	}

	return "partial"
}

// buildReflectionPrompt constructs the user message for the reflection
// agent.
func buildReflectionPrompt(analysis SessionAnalysis, messages []Message) string {
	var b strings.Builder

	b.WriteString("## Session analysis\n\n")

	if analysis.TaskDescription != "" {
		b.WriteString(fmt.Sprintf("**Task:** %s\n\n", analysis.TaskDescription))
	}

	if analysis.ToolCallSummary != "" {
		b.WriteString(fmt.Sprintf("**Tool usage:**\n%s\n\n", analysis.ToolCallSummary))
	}

	b.WriteString(fmt.Sprintf("**Iterations:** %d tool-call rounds\n", analysis.IterationCount))
	b.WriteString(fmt.Sprintf("**Write/edit calls:** %d\n", analysis.WriteToolCalls))
	b.WriteString(fmt.Sprintf("**Inferred outcome:** %s\n\n", analysis.Outcome))

	if len(analysis.ErrorsEncountered) > 0 {
		b.WriteString("**Errors encountered:**\n")
		for _, e := range analysis.ErrorsEncountered {
			b.WriteString(fmt.Sprintf("- %s\n", e))
		}
		b.WriteString("\n")
	}

	if len(analysis.UserCorrections) > 0 {
		b.WriteString("**User corrections:**\n")
		for _, c := range analysis.UserCorrections {
			b.WriteString(fmt.Sprintf("- %s\n", c))
		}
		b.WriteString("\n")
	}

	b.WriteString("## Conversation transcript\n\n")
	for _, m := range messages {
		switch m.Role {
		case RoleUser:
			content := m.Content
			if len(content) > 1000 {
				content = content[:1000] + "\n[...truncated]"
			}
			b.WriteString(fmt.Sprintf("[user]: %s\n\n", content))

		case RoleAssistant:
			content := m.Content
			if len(content) > 1000 {
				content = content[:1000] + "\n[...truncated]"
			}
			if content != "" {
				b.WriteString(fmt.Sprintf("[assistant]: %s\n\n", content))
			}
			for _, tc := range m.ToolCalls {
				args := string(tc.Arguments)
				if len(args) > 200 {
					args = args[:200] + "..."
				}
				b.WriteString(fmt.Sprintf("[tool_call %s]: %s\n\n", tc.Name, args))
			}

		case RoleTool:
			content := m.Content
			if len(content) > 300 {
				content = content[:300] + "..."
			}
			b.WriteString(fmt.Sprintf("[tool_result %s]: %s\n\n", m.Name, content))
		}
	}

	return b.String()
}

// parseReflectionResult extracts the reflection data from the model's
// JSON response.
func parseReflectionResult(content string) ReflectionResult {
	content = strings.TrimSpace(content)

	if idx := strings.Index(content, "{"); idx >= 0 {
		if end := strings.LastIndex(content, "}"); end > idx {
			content = content[idx : end+1]
		}
	}

	var result ReflectionResult
	if err := json.Unmarshal([]byte(content), &result); err != nil {
		return ReflectionResult{}
	}

	return result
}
